datalad/interface/unlock.py

```diff
@@ -69,97 +72,76 @@ def __call__(
-                    ap['message'] = "not annex'ed, nothing to unlock"
-                    ap.update(res_kwargs)
-                    yield ap
-                continue
-
-            # only files in annex with their content present:
-            files = [ap['path'] for ap in content]
-            to_unlock = []
-            for ap, under_annex, has_content in \
-                zip(content,
-                    ds.repo.is_under_annex(files),
-                    ds.repo.file_has_content(files)):
-
-                # TODO: what about directories? Make sure, there is no
-                # situation like no file beneath with content or everything in
-                # git, that leads to a CommandError
-                # For now pass to annex:
-                from os.path import isdir
-                if isdir(ap['path']):
-                    to_unlock.append(ap)
-                    continue
-
-                # Note, that `file_has_content` is (planned to report) True on
-                # files in git. Therefore order matters: First check for annex!
-                if under_annex:
-                    if has_content:
-                        to_unlock.append(ap)
-                    # no content, no unlock:
-                    else:
-                        ap['status'] = 'impossible'
-                        ap['message'] = "no content present, can't unlock"
-                        ap.update(res_kwargs)
-                        yield ap
-                # file in git, no unlock:
+            has_content = res.get("has_content")
+            if has_content:
+                to_unlock[res["parentds"]].append(res["path"])
+            elif paths_nondir and Path(res["path"]) in paths_nondir:
+                if has_content is False:
+                    msg = "no content present"
+                elif res["state"] == "untracked":
+                    msg = "untracked"
                 else:
-                    ap['status'] = 'notneeded'
-                    ap['message'] = "not controlled by annex, nothing to unlock"
-                    ap.update(res_kwargs)
-                    yield ap
-
-            # don't call annex-unlock with no path, if this is this case because
-            # nothing survived the filtering above
-            if content and not to_unlock:
-                continue
+                    # TODO: Anyway to distinguish between a regular git file
+                    # and an unlocked annex file?
+                    msg = "non-annex file"
+                yield get_status_dict(
+                    status="impossible",
+                    path=res["path"],
+                    type="file",
+                    message="{}; cannot unlock".format(msg),
+                    **res_kwargs)
 
-            for r in ds.repo.unlock([ap['path'] for ap in to_unlock]):
+        # Do the actual unlocking.
+        for ds_path, files in to_unlock.items():
+            ds = Dataset(ds_path)
```

Comment:
> We should check for the dataset being an `AnnexRepo` to avoid a crash

These are only repos for files that had `has_content`, thus annex files, so I think we're safe in assuming it's an AnnexRepo at this point.  That being said, I don't mind adding an additional check.

> , and issue `notneeded` results. `notneeded` instead of `impossible`, because we want to focus on the result, and not on the process for max sanity of user feedback.

In the current code's state, this applies to the code above (in the `elif` arm of the `has_content` condition).  I went with `impossible` just because that is the current behavior, but I agree the `notneeded` makes more sense for already unlocked files or regular git files.  I'll update to it to do so.

Reply:
me:
> I went with `impossible` just because that is the current behavior

Actually unlock's current behavior is to yield "notneeded" results.  I don't know how I got mixed about that.